### PR TITLE
Remove thread check dependency on metrics [PLEN-96]

### PR DESCRIPTION
**Pull request by nicu-da**
Thread checks now accept a function that returns the queue size instead of the metrics and using the metrics to calculate the queue size.
As a performance optimization we added a wrapper around the executors, that keeps track of the queue size. This is done to avoid reading the queue size directly from the executor, as it's not a simple operation as explained here https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ForkJoinPool.html#getQueuedSubmissionCount--

### CODE REVIEW

File: ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/SandboxOnXRunner.scala [hunk 1]
```diff
@@ -230,7 +232,7 @@ object SandboxOnXRunner {
   private def buildServicesExecutionContext(
       metrics: Metrics,
       servicesThreadPoolSize: Int,
-  ): ResourceOwner[ExecutionContextExecutorService] =
+  ): ResourceOwner[QueueAwareExecutionContextExecutorService] =
     ResourceOwner
```
Comment (simonmaxen-da) [inline ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/SandboxOnXRunner.scala:236]: This binds the metric name to the executor service. This name is then re-associated on line 297.  I don't like that, can we either:
- Add a metric name to QueueAwareExecutionContextExecutorService
- Return a tuple (QueueAwareExecutionContextExecutorService, MetricName)

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/ratelimiting/ThreadpoolCheck.scala [hunk 2]
```diff
@@ -19,33 +16,23 @@ object ThreadpoolCheck {
   private implicit val logger: ContextualizedErrorLogger =
     DamlContextualizedErrorLogger.forClass(getClass)
 
-  /** Match naming in [[com.codahale.metrics.InstrumentedExecutorService]] */
```
Comment (simonmaxen-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/ratelimiting/ThreadpoolCheck.scala:22]: Nice to see this going 👍

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/ratelimiting/RateLimitingInterceptor.scala [hunk 3]
```diff
@@ -83,13 +83,15 @@ object RateLimitingInterceptor {
       metrics: Metrics,
       config: RateLimitingConfig,
       additionalChecks: List[LimitResultCheck] = List.empty,
+      indexDbQueueSizeProvider: () => Long,
```
Comment (simonmaxen-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/ratelimiting/RateLimitingInterceptor.scala:86]: We don't need this if we wrap the index DB queue size check in a `LimitResultCheck`

File: ledger/ledger-api-errors/src/main/scala/com/daml/error/definitions/LedgerApiErrors.scala [hunk 4]
```diff
@@ -125,18 +126,19 @@ object LedgerApiErrors extends LedgerApiErrorGroup {
...
         name: String,
+        metricNameLabel: String,
         queued: Long,
         limit: Int,
-        metricPrefix: String,
         fullMethodName: String,
     )(implicit errorLogger: ContextualizedErrorLogger)
         extends DamlErrorWithDefiniteAnswer(
-          s"The $name queue size ($queued) has exceeded the maximum ($limit). Api services metrics are available at $metricPrefix.",
+          s"The $name queue size ($queued) has exceeded the maximum ($limit). Metrics for queue size available at ${ExecutorServiceMetrics.CommonMetricsName.QueuedTasks}.",
```
Comment (simonmaxen-da) [inline ledger/ledger-api-errors/src/main/scala/com/daml/error/definitions/LedgerApiErrors.scala:135]: There seems to be two sources of metrics naming `ExecutorServiceMetrics` and `com.daml.metrics.Metrics`

File: ledger/participant-integration-api/src/main/scala/platform/LedgerApiServer.scala [hunk 5]
```diff
@@ -50,7 +51,9 @@ class LedgerApiServer(
     //          Currently, we provide this flag outside the HOCON configuration objects
     //          in order to ensure that participants cannot be configured to accept explicitly disclosed contracts.
     explicitDisclosureUnsafeEnabled: Boolean = false,
-    rateLimitingInterceptor: Option[RateLimitingInterceptor] = None,
+    rateLimitingInterceptor: Option[
+      QueueAwareExecutionContextExecutorService => RateLimitingInterceptor
```
Comment (simonmaxen-da) [inline ledger/participant-integration-api/src/main/scala/platform/LedgerApiServer.scala:55]: Seems strange, the executor service is created outside the LedgerApiServer.  Seems cleaner to make the association outside of the Ledger Api Server (as now).

File: ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/SandboxOnXRunner.scala (hunk 1 above)
Comment (nicu-da) [reply ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/SandboxOnXRunner.scala:236]: Good point, will go with the first approach as it's simpler and cleaner.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/ratelimiting/RateLimitingInterceptor.scala (hunk 3 above)
Comment (nicu-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/ratelimiting/RateLimitingInterceptor.scala:86]: How would this work? We need a way to read the current queue size

File: ledger/ledger-api-errors/src/main/scala/com/daml/error/definitions/LedgerApiErrors.scala (hunk 4 above)
Comment (nicu-da) [reply ledger/ledger-api-errors/src/main/scala/com/daml/error/definitions/LedgerApiErrors.scala:135]: `ExecutorServiceMetrics` are specific to executors. We in fact create the `ExecutorServiceMetrics` instance in `com.daml.metrics.Metrics`[

File: ledger/participant-integration-api/src/main/scala/platform/LedgerApiServer.scala (hunk 5 above)
Comment (nicu-da) [reply ledger/participant-integration-api/src/main/scala/platform/LedgerApiServer.scala:55]: The problem is that the executor service is actually created in the `Indexer`, which is created as part of the `LedgerApiServer`. This is why we need to pass a way to create an interceptor from an existing executor service.
To refactor this we would need to refactor the start-up phase or move the executor service creation for the indexer a few levels up. I found this change to be the less disruptive from all of them.

Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/LedgerApiServer.scala:55]: Yes.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/ratelimiting/RateLimitingInterceptor.scala (hunk 3 above)
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/ratelimiting/RateLimitingInterceptor.scala:86]: I would do it in SandboxOnXRunner.buildRateLimitingInterceptor

File: ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/SandboxOnXRunner.scala [hunk 6]
```diff
@@ -285,16 +287,24 @@ object SandboxOnXRunner {
   }
 
   def buildRateLimitingInterceptor(
```
Comment (simonmaxen-da) [inline ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/SandboxOnXRunner.scala:289]: I would do:
```
  def buildRateLimitingInterceptor(
      metrics: Metrics,
      indexDbExecutor: QueueAwareExecutionContextExecutorService,
      apiServicesExecutor: QueueAwareExecutionContextExecutorService,
  )(config: RateLimitingConfig): RateLimitingInterceptor = {

    val indexDbCheck = ThreadpoolCheck(
      "Index DB Threadpool",
      indexDbExecutor,
      config.maxApiServicesIndexDbQueueSize,
    )

    val apiServicesCheck = ThreadpoolCheck(
      "Api Services Threadpool",
      apiServicesExecutor,
      config.maxApiServicesQueueSize,
    )

    RateLimitingInterceptor(metrics, config, List(indexDbCheck, apiServicesCheck))

  }
```

Comment (simonmaxen-da) [reply ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/SandboxOnXRunner.scala:289]: We will also have to fix this up on the Canton side, but this is fine as they create their own threadpools so it makes sense for them to create the threadpool checks (as is currently the case).

Comment (nicu-da) [reply ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/SandboxOnXRunner.scala:289]: Ah alright, I'm much in favour of this approach. Chose not to go down that path as I assumed we want to enforce the indexer thread pool check.

**Comment by nicu-da** [comment]
@simonmaxen-da thank you for the review and the changes removing the explicit index db check, all ready for re-review

**Review by simonmaxen-da — APPROVED** [review]
Looks good on the participant side. we need to make a note to add the indexer db threadpool check on the Canton side.

**Comment by nicu-da** [comment]
> Looks good on the participant side. we need to make a note to add the indexer db threadpool check on the Canton side.

@simonmaxen-da Yes, once this is merged I will wrap up the canton side implementation and open the PR there as well.

File: docs/resources/generated-error-pages/error-codes-inventory.rst.inc [hunk 7]
```diff
@@ -456,7 +456,7 @@ THREADPOOL_OVERLOADED
 
     **Conveyance**: This error is logged with log-level INFO on the server side and exposed on the API with grpc-status ABORTED including a detailed error message.
 
-    **Resolution**: The following actions can be taken: Here the 'queue size' for the threadpool = 'submitted tasks' - 'completed tasks' - 'running tasks' 1. Review the historical 'queue size' growth by inspecting the metric given in the message. 2. Review the maximum 'queue size' limits configured in the rate limiting configuration. 3. Try to space out requests that are likely to require a lot of CPU or database power.      
+    **Resolution**: The following actions can be taken: Here the 'queue size' for the threadpool is considered as reported by the executor itself. 1. Review the historical 'queue size' growth by inspecting the metric given in the message. 2. Review the maximum 'queue size' limits configured in the rate limiting configuration. 3. Try to space out requests that are likely to require a lot of CPU or database power.
```
Comment (da-katmurp) [inline docs/resources/generated-error-pages/error-codes-inventory.rst.inc:459]: Can you make this a numbered list?

**Review by da-katmurp — APPROVED** [review]

**Pull request merged**